common/routing/route.go

```diff
@@ -0,0 +1,140 @@
+type Route[T any] struct {
+	components []Component[T]
+}
+
+// Component represents a single HTTP path component, either a constant slug or a variable parameter.
+type Component[T any] interface {
+	// Representation is the string representation of the component for usage in a path definition, e.g. "v1" for a
+	// constant slug or "{namespace}" for a variable. This should be compatible with the format specified in
+	// the [github.com/gorilla/mux] package.
+	Representation() string
+	// Serialize returns the actual value of the slug when used in an HTTP path, e.g. "v1" for a constant slug or
+	// "test-namespace" for a variable.
+	Serialize(params *T) string
+	// Deserialize mutates the given params object with the value of the component when parsing an HTTP path, e.g.
+	// setting the value of a variable to "test-namespace". If the component is a constant slug, this method should
+	// be a no-op.
+	Deserialize(vars map[string]string, t *T)
+}
+
+// NewRoute creates a new Route with the given options.
+func NewRoute[T any](components ...Component[T]) Route[T] {
+	return Route[T]{components}
+}
+
+// Representation returns the [github.com/gorilla/mux] compatible string representation of the route for usage in a
+// path definition. It does not add a leading or trailing slash to the representation, but it won't remove them if
+// they're present in the components. We do this because it's easier to add a slash depending on the context than to
+// remove it.
+func (r Route[T]) Representation() string {
+	return r.serialize(func(c Component[T]) string {
+		return c.Representation()
+	})
+}
+
+// Path returns the serialized path of the route with the given params. There will be no leading or trailing slashes,
+// similar to the behavior of the [Route.Representation] method.
+func (r Route[T]) Path(t *T) string {
+	return r.serialize(func(c Component[T]) string {
+		return c.Serialize(t)
+	})
+}
+
+func (r Route[T]) serialize(f func(c Component[T]) string) string {
+	var sb strings.Builder
+	for i, c := range r.components {
+		if i > 0 {
+			sb.WriteString("/")
+		}
+		sb.WriteString(f(c))
+	}
+	return sb.String()
+}
+
+// Deserialize the given vars into a new instance of the params type, T.
+func (r Route[T]) Deserialize(vars map[string]string) *T {
+	var t T
+	for _, c := range r.components {
+		c.Deserialize(vars, &t)
+	}
+	return &t
+}
+
+// Slugs returns a [Component] that represents a series of HTTP path slugs in a Route. Don't use slashes in the slugs.
+func Slugs[T any](value ...string) slugs[T] {
+	return value
+}
+
+type slugs[T any] []string
+
+func (s slugs[T]) Representation() string {
+	return strings.Join(s, "/")
+}
+
+func (s slugs[T]) Serialize(*T) string {
+	return s.Representation()
+}
+
+func (s slugs[T]) Deserialize(map[string]string, *T) {}
+
+// StringParam returns a [Component] that represents a string variable in a Route.
+func StringParam[T any](name string, getter func(*T) *string) stringParam[T] {
```

Comment:
Maybe want to also support pattern but we can add that when we have a use case.

Reply:
Yeah, I think we definitely could want that later, especially since mux supports it.